model/mobile_trigger.go

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"io"
+)
+
+const (
+	ScopeMobile = "mobile"
+	ScopeWebApp = "webapp"
+)
+
+type PluginIntegration struct {
+	PluginID   string   `json:"id"`
+	Location   string   `json:"location"`
+	RequestURL string   `json:"request_url"`
+	Scope      []string `json:"scope"`
+	//	Extra    interface{} `json:"extra"`
+	Extra string `json:"extra"` // Should use interface{} to be able to create different extras for different locations
+}
+
+type MobileIntegrationChannelHeader struct {
+	DefaultMessage string `json:"default_message"`
```

Comment:
How are these specific location structs intended to be used? Is this to be set in the `Extra` field?

I believe it would be simpler to initially add the `text` (and `icon`) directly to the plugin action struct that would probably be commonly used for most  action and once we have more usage we can think of abstracting into this or a concept of custom metadata.
It would also simplify at the client level of not having to special case handling to process the different type of data that the plugin action definition might have for something like text that most of them will use.

Reply:
Yes, I am having some problems with RPC, and that is why it is not implemented as intended.
The main idea is to have for each "integration" (I should have changed the name of this one) a struct to pass to extra. For channel header, it could be something like:
- Icon
- Text
- HasToggle
- CountURL

Meanwhile, for post menu would be just
- Icon
- Text

And on one of the most ambitious cases, a "RHS view" would be:
- HeaderTitle
- RootComponent

I know many will have text and icon, but not all of them will have, and I want to keep the open mind.

Reply:
I like @marianunez's idea of starting with simple `text` and `icon` fields inside of the main `struct` and iterate on it once we now the other use cases. `text` and `icon` will cover most cases for now. We can still add an `Extra` field later, if we need it. And it fixes the RPC problems.

Reply:
As discussed offline, I want to keep the doors open to any possible location. There are several locations considered on the webapp, and I would like the possibility to reflect that with this. Here is the list if I am not mistaken:
- Root Component
- Popover User Attributes
- Popover User Actions
- Left Sidebar Header
- Bottom Team Sidebar
- Post Message Attachment
- Link Tooltip
- Channel Header Button
- Post Type
- Post Card Type
- Main Menu Action
- Channel Header Menu Action
- Post Dropdown Menu Action (and submenus)
- File Upload Method
- Admin Console Custom Setting
- RHS Component
- File preview

And also there are many "actions" that can be considered a "plugin integration".
- File will upload
- Websockets events
- Reconnect handler
- Message will be posted
- Slash command will be posted
- Message will format

Some of them are more easy to implement on this context, some are more complicated. We can decide on just a few to do the first version (the most used, the easiest, or anything like that), but as you see, not all would need a text or an icon.
I don't want to over-scope this functionality, but I want to keep the door open to extend it if needed to cover all the cases.